common/lib/plugins/bluegreen/blue_green_status_provider.ts

```diff
@@ -0,0 +1,939 @@
+    this.statusCheckIntervalMap.set(BlueGreenIntervalRate.INCREASED, BigInt(WrapperProperties.BG_INTERVAL_INCREASED_MS.get(properties)));
+    this.statusCheckIntervalMap.set(BlueGreenIntervalRate.HIGH, BigInt(WrapperProperties.BG_INTERVAL_HIGH_MS.get(properties)));
+
+    this.switchoverTimeoutNanos = convertMsToNanos(WrapperProperties.BG_SWITCHOVER_TIMEOUT_MS.get(properties));
+    this.suspendNewBlueConnectionsWhenInProgress = WrapperProperties.BG_SUSPEND_NEW_BLUE_CONNECTIONS.get(properties);
+
+    const dialect: DatabaseDialect = this.pluginService.getDialect();
+    if (this.isBlueGreenDialect(dialect)) {
+      this.initMonitoring();
+    } else {
+      logger.warn(Messages.get("Bgd.unsupportedDialect", this.bgdId, dialect.getDialectName()));
+    }
+  }
+
+  protected initMonitoring(): void {
+    this.monitors[BlueGreenRole.SOURCE.value] = new BlueGreenStatusMonitor(
+      BlueGreenRole.SOURCE,
+      this.bgdId,
+      this.pluginService.getCurrentHostInfo(),
+      this.pluginService,
+      this.getMonitoringProperties(),
+      this.statusCheckIntervalMap,
+      { onBlueGreenStatusChanged: (role, status) => this.prepareStatus(role, status) }
+    );
+
+    this.monitors[BlueGreenRole.TARGET.value] = new BlueGreenStatusMonitor(
+      BlueGreenRole.TARGET,
+      this.bgdId,
+      this.pluginService.getCurrentHostInfo(),
+      this.pluginService,
+      this.getMonitoringProperties(),
+      this.statusCheckIntervalMap,
+      { onBlueGreenStatusChanged: (role, status) => this.prepareStatus(role, status) }
+    );
+  }
+
+  protected getMonitoringProperties(): Map<string, any> {
+    const monitoringConnProperties: Map<string, any> = new Map(this.properties);
+
+    for (const key of monitoringConnProperties.keys()) {
+      if (!key.startsWith(BlueGreenStatusProvider.MONITORING_PROPERTY_PREFIX)) {
+        continue;
+      }
+
+      monitoringConnProperties.delete(key);
+    }
+
+    if (!monitoringConnProperties.has(WrapperProperties.WRAPPER_CONNECT_TIMEOUT.name)) {
+      WrapperProperties.WRAPPER_CONNECT_TIMEOUT.set(monitoringConnProperties, BlueGreenStatusProvider.DEFAULT_CONNECT_TIMEOUT_MS);
+    }
+
+    if (!monitoringConnProperties.has(WrapperProperties.WRAPPER_QUERY_TIMEOUT.name)) {
+      WrapperProperties.WRAPPER_QUERY_TIMEOUT.set(monitoringConnProperties, BlueGreenStatusProvider.DEFAULT_QUERY_TIMEOUT_MS);
+    }
+
+    return monitoringConnProperties;
+  }
+
+  protected isBlueGreenDialect(dialect: any): dialect is BlueGreenDialect {
+    return dialect;
+  }
+
+  protected async prepareStatus(role: BlueGreenRole, interimStatus: BlueGreenInterimStatus): Promise<void> {
+    // Detect changes
+    const contextHash: number = this.getContextHash();
+
+    const storedStatus = this.interimStatuses[role.value];
+
+    if (_.isEqual(storedStatus, interimStatus) && this.lastContextHash === contextHash) {
```

Comment:
In JDBC we compare hashes for equality instead of the checking storedStatus and interimStatus directly, do we need to do the same here?

Reply:
_.isEqual is comparing the contents of storedStatus and interimStatus, I believe it achieves the same goal as comparing the custom hashcode.